src/parser.rs

```diff
@@ -1311,20 +1324,45 @@ impl Library {
             _ => Err(parser.unexpected_element(elem)),
         })?;
 
-        if let Some((tid, c_type, mut array_length)) = typ {
-            if for_method {
-                array_length = array_length.map(|l| l + 1);
+        if let Some(mut elem_type) = typ {
+            if elem_type.is_array {
+                // Note: these could be handled in the `analysis` part, however that would be quite
+                //       annoying because many code places check the parsed `nullable` instead of
+                //       the `analysis` result. Gir necessitates a rewrite...
+                use ParameterDirection::*;
+                match direction {
+                    Return | Out | InOut => {
+                        // Some arrays in return position come with nullable=true & zero-terminated=false
+                        // annotations. In practice, the C implementations return NULL when the size is 0.
+                        // So it doesn't make much sense using an `Option<_>` by default for nullable arrays.
+                        nullable = false
+                    }
+                    _ => {
+                        // We can only generate a Rust `Option` from a nullable array annotation
+                        // if the array is not zero terminated. Otherwise we would need an extra
+                        // indication to differentiate an empty array from an undefined value,
+                        // which should be indicated in the doc and would need manual implementation.
```

Comment:
I don't understand this part. With `zero-terminated=true` it actually seems clear how to handle nullability but not otherwise (see above comment)

Reply:
Yes, I'm not sure what led me to this comment & condition... I must have been too sleepy :sweat_smile: 
Updated and also updated the [`gtk-rs-core`](https://github.com/gtk-rs/gtk-rs-core/pull/1728) & [`gtk4`](https://github.com/gtk-rs/gtk4-rs/pull/2064) PRs.